### ISSUE TITLE
Release dates for PocketMine-MP 4.0.0

### ISSUE THREAD

**Issue by dktapps** [pmmp/PocketMine-MP#4561]
It's time to get testing, plugin developers!

## Release dates for PocketMine-MP 4.0.0
PM4 is currently targeting a release date of **December 1st, 2021**.
This is **not set in stone** and may be delayed if new major bugs appear or if a MC update comes along to cause headaches.

## End-of-support dates for PocketMine-MP 3.x
After 4.0.0 release, PM3 will receive Minecraft version updates and critical bug fixes only until the end-of-life (EOL) date.
The EOL date currently planned for PM3 is on or after **March 1st, 2022**.
Again, this is **not set in stone** and could be extended or delayed depending on what happens.

This gives y'all **3 months after 4.0.0 full release** to update your plugins to PM4 and get upgraded.

**Comment by sn3akrr** [comment]
Hallelujah

**Comment by redmcteam** [comment]
epic

**Comment by efor89** [comment]
mhh i think the release date come to fast pmmp 4 missing so much items, blocks and a lot of functions of blocks and items and if i lock on issues.... ^^ plz dont release a half rdy pmmp4

**Comment by MonoAdrian23** [comment]
Use Powernukkit

**Comment by dktapps** [comment]
> mhh i think the release date come to fast pmmp 4 missing so much items, blocks and a lot of functions of blocks and items and if i lock on issues.... ^^ plz dont release a half rdy pmmp4

PM4's goal was never to implement vanilla features. Those aren't considered essential for a final release.

**Comment by efor89** [comment]
ok but then i dont know why i need pmmp4? 
wat is new in pmmp4 that the blocks have not a negative id and better peformance.
with pmmp3 its all ok it work good i have all items and blocks 1.13, 1.14, 1.15, 1.16 and some items of 1.17
and i can catch fish and so on.

**Comment by redmcteam** [comment]
> ok but then i dont know why i need pmmp4?
> wat is new in pmmp4 that the blocks have not a negative id and better peformance.
> with pmmp3 its all ok it work good i have all items and blocks 1.13, 1.14, 1.15, 1.16 and some items of 1.17
> and i can catch fish and so on.

anyone can implement new blocks, items, mob AI and mobs by creating a plugin.

**Comment by efor89** [comment]
yes but i have it in pmmp3 so why i need pmmp4?

**Comment by redmcteam** [comment]
> yes but i have it in pmmp3 so why i need pmmp4?

you need pmmp4 for implement new 1.13+ blocks, pmmp3 not supports blocks with negative ID

**Comment by efor89** [comment]
extendedblocks work good

**Comment by redmcteam** [comment]
> extendedblocks work good

ExtendedBlocks is not good solution.

**Comment by ColinHDev** [comment]
You need pm4, cause pm3 will be discontinued in March

**Comment by dktapps** [comment]
You can read about what's changed here: https://github.com/pmmp/PocketMine-MP/blob/master/changelogs/4.0.md

The TL;DR is that there are a large number of improvements all around, performance, memory usage, plugin API improvements, world format support, IPv6 and many other things.

But the biggest reason is, like Colin said, that PM3 will no longer receive any updates after March 2022.

**Comment by FlxiBoy1313** [comment]
Dylan, will there really be all blocks for PM4, or will there be 1.12 blocks again?

**Comment by dktapps** [comment]
Currently only blocks up to 1.12 are supported by PM itself, but with the newer world format support it's much easier for plugins to add their own without destroying server performance (read: ExtendedBlocks).

The reason 1.13+ is not yet supported is because of a second change in the world format, which is an extremely large change.

I _could_ implement 1.13+ blocks on PM3 by extending the old format with custom id/metadata (the 1.12 format has space for infinite blocks, it just doesn't support NBT blockstate), however I chose not to do that for the following reasons:

- Adding custom things to a standard format makes it hard for third-party tools like MCEdit (and MC itself) to operate on worlds. Anyone who was around when PMAnvil was added knows why this is a bad idea.
- Adding custom things directly increases the work that I, as a maintainer, have to do. Any custom thing would have to have conversion code written for it to be supported once the new format is actually implemented.

The bottom line is that 1.13+ blocks will not be supported by PM itself until the newer world format support is implemented. This will not be in PM4 since it's many years overdue already.

**Comment by dktapps** [comment]
I should also point out that if you wanted a vanilla experience, you came to the wrong place to begin with, and should be using BDS.

**Comment by FlxiBoy1313** [comment]
But that's not a solution, is it? You have to use plugins for everything?

**Comment by efor89** [comment]
i know that pmmp4 is better than 3 and thats the reason i donate pmmp.
but wat i see is a half rdy pmmp4 version.

**Comment by dktapps** [comment]
You see that because you don't understand the point of PM4. Like I said, PM4 was never about vanilla features.

**Comment by dktapps** [comment]
I'm going to lock this issue so it stops noising people's notifications. If you want to continue this debate, please take it to our [Discord server](https://discord.gg/bmSAZBG) or [forums](https://forums.pmmp.io).
